### PR TITLE
fixed issues with relation data read

### LINKED ISSUE #784: Better error messages on permission errors with accessing relation databags
In non-peer relations, a unit cannot read or write its own application's databag, which results into authorization errors.
However, those errors surface in unexpected places which can confuse developers:

```
self.unit.is_leader() # False
self.relation.data # -->  ModelError("ERROR permission denied")
self.relation.data[relation.app]  # {}
self.relation.data[self.app]  # -->  ModelError("ERROR permission denied")
self.relation.data[relation.app]['foo'] = 'bar' # -->  ModelError("ERROR permission denied")
self.relation.data[relation.units[0]]['foo'] = 'bar' # -->  ModelError("ERROR permission denied") 
```

goal:
```
self.unit.is_leader() # False
self.relation.data # RelationData(local_app: <N/A>, local_unit: {...}, remote_app: {...}, remote_unit: {...}) 
self.relation.data[relation.app]  # {}
self.relation.data[self.app]  # -->  ModelError("ERROR permission denied: non-leader units cannot read or write their own application's databag.")
self.relation.data[relation.app]['foo'] = 'bar' # -->  ModelError("ERROR permission denied: cannot write databags of remote applications")
self.relation.data[relation.units[0]]['foo'] = 'bar' # -->  ModelError("ERROR permission denied: cannot write databags of remote units"") 
```

### PR DESCRIPTION
**Pull request by PietroPasotti**
With these changes:

- `repr(relation[self.app])` will be "<n/a>" if you're not the leader, no longer causing errors
- You get a different error message if you attempt to access keys from that databag (vs. if you try to write to it)
  - In order to do that, I added a `RelationDataContents._is_readable` attr. For symmetry, I renamed `_is_mutable` to `_is_writeable`. Also updated the harness' counterparts with similar plumbing.
- `repr(relation.data)` should not raise exceptions, in real ops and in harness, but the representations should also match (in strict mode at least: see below).
- the harness' backend is now access-controlling databags as the real backend would.
  (note): for that to be possible, while still allowing test code to have unrestricted read/write access to the databags, we rely on the _is_hook_running flag set on the backend.
- In order to allow test code to enable 'strict' checking mode even when one is not relying on the Harness' internal machinery (_reemit specifically), I promoted EventContext to a `Harness.backend.event_context` public method (also directly accessible as `Harness.event_context`).
- tests can still read/write at will databags via the harness' `[update/get]_relation_data` methods; these have been updated to use `event_context` to ensure that strict mode is disabled.

I think this is reasonable because otherwise it'd be too hard to switch between the two harness operation modes (strict: do as if we're really inside a charm; and admin: we're mocking and testing stuff).

## Checklist

 - [x] Have any types changed? If so, have the type annotations been updated?
 - [x] If this code exposes model data, does it do so thoughtfully, in a way that aids understanding?
 - [x] Do error messages, if any, present charm authors or operators with enough information to solve the problem?

## QA steps

Added unittests to check that model behaviour matches the expectations.

## Bug reference

Fixes #784

## Changelog

- `ops` gives better error messages when charms attempt to interact with relation databags they don't have access to.

### CODE REVIEW

**Comment by PietroPasotti** [comment]
We need to discuss: harness testing model backend serves at the moment two purposes:
1) Mock the interactions from the charm's POV (restricted by the same rule as a production charm would be)
2) Offer the testing environment tools to read/write databags (unrestrictedly)

These two needs are now starting to clash because we want the harness model backend to behave as juju does when a charm tries to read/write the data, but to leave tests 'admin access' when they are doing their testing.

How do we solve this?

Proposal: 
relation objects facing the charm return 'locked' (mock) RelationDataContents objects,
relation objects facing the harness return 'unlocked' ones.

Therefore:
```python
harness.set_leader(False)
harness.get_relation('foo').data[harness.charm.app]['foo'] = 'bar'  # ok

harness.charm.model.get_relation('foo').data[harness.charm.app]['foo'] = 'bar'  # ERROR!
```

@pengale @rwcarlsen @jnsgruk @jameinel

**Comment by PietroPasotti** [comment]
Robert recommended to check the global env var 'am I inside an event hook' and only if YES then raise an error, else keep it quiet and behave as if we're in 'admin mode'.

Will do that and think of any edge cases we might be hitting (if someone is 'manually' emitting events by calling handlers with mocked events for example...) and mitigating strategies

**Comment by PietroPasotti** [comment]
In conclusion:
- the harness' backend is now access-controlling databags as the real backend would.
- for that to be possible, while still allowing test code to have unrestricted read/write access to the databags, we rely on the _is_hook_running flag set on the backend.
- In order to allow test code to enable 'strict' checking mode even when one is not relying on the Harness' internal machinery (_reemit specifically), I promoted EventContext to a `Harness.backend.event_context` public method (also directly accessible as `Harness.event_context`).

I think this is reasonable because otherwise it'd be too hard to switch between the two harness operation modes (strict: do as if we're really inside a charm; and admin: we're mocking and testing stuff).

**Comment by PietroPasotti** [comment]
RE the failing test: there's an assertBackendCalls that breaks for funny reasons.
It's the backend calls generated by `repr(relation.data)`

in python > 3.5 we get:
```
    [('relation_get', 1, 'myapp/0', False),
     ('is_leader',),
     ('relation_get', 1, 'remoteapp1/0', False),
     ('relation_get', 1, 'remoteapp1', True)]
     ```

in python 3.5 for some reason we get instead:
```
    [('relation_get', 1, 'remoteapp1/0', False),
     ('relation_get', 1, 'remoteapp1', True),
     ('is_leader',),
     ('relation_get', 1, 'myapp/0', False)]
     ```

I think it's got something to do with dict sorting :D

File: ops/model.py
Comment (rwcarlsen) [inline ops/model.py:951]: Maybe check for the opposite side of this - e.g. check for `if i_am_remote or i_am_leader: return True` - seems like it could clean up and simplify this logic.

File: ops/model.py
Comment (rwcarlsen) [inline ops/model.py:917]: Semantically, this name seems like it should be something more like "_hook_is_running".

File: ops/model.py
Comment (rwcarlsen) [inline ops/model.py:976]: Feels like this error message text should be consistent with the get/read one below.

File: ops/testing.py
Comment (rwcarlsen) [reply ops/testing.py:940]: You shouldn't get any restrictions if a hook isn't running.  It seems to me like the admin access only matters when a hook isn't running - e.g. in test setup or test assertions.

File: ops/model.py
Comment (PietroPasotti) [reply ops/model.py:917]: fair enough, renaming the underlying call seems bad

File: ops/testing.py
Comment (PietroPasotti) [reply ops/testing.py:820]: It would be unnecessary if the user were running that call from test code, normally.
However, suppose that the user calls `harness.get_relation_data` while inside a debugging session (e.g. put a breakpoint in `Charm._on_pebble_ready`), so while a hook IS in fact running. Then if the user were to do `harness.set_relation_data`, it would fail because you're inside a hook context).
My reasoning was: we want to ensure that `harness.set_relation_data` ALWAYS works the same, i.e. bypasses all checks.
Hence the `event_context('')`: this tells the harness: do this call always as if no hook were running.

File: ops/testing.py
Comment (PietroPasotti) [reply ops/testing.py:141]: I agree for 90% of the cases, but I could think of a few scenarios:

The most common one: I want to test an event handler but mock the event, so without going through `emit()`. So I want my test code to be:
```python
def test_evt_handler():
    evt = MagicMock()
    evt.attr_I_cannot_fake.in.dev.return_value = 42
    MyCharm()._on_fancy_event(evt)
```

How can I ensure that the harness be strict in validating my relation data access in that event handler?
```python
def test_evt_handler():
    evt = MagicMock()
    evt.attr_I_cannot_fake.in.dev.return_value = 42
    harness = Harness(MyCharm)
    with harness.event_context('fancy_event'):
        harness.charm._on_fancy_event(evt)
```

File: ops/testing.py
Comment (PietroPasotti) [reply ops/testing.py:623]: ach, so bad... 
I went a few times back and forth with that logic because I got jumbled several times with keeping in mind who could read what. Good catch!

File: ops/testing.py
Comment (PietroPasotti) [reply ops/testing.py:637]: You're right. It's much cleaner. I'll fix that

File: ops/testing.py
Comment (PietroPasotti) [reply ops/testing.py:820]: It becomes redundant now that I separate the raw data in the backend and use RelationDataContent to access it.

File: ops/testing.py
Comment (PietroPasotti) [reply ops/testing.py:942]: right, but the _hook_is_running thing is private, and with the newer 'strict' access control, we need IMHO to expose the finer knobs to the user. 
Two cases:
the user is inside hook code, and wants to do stuff 'as if' it was outside a hook context
  - less common, could happen when setting a breakpoint inside a tester charm's hooks
the user is outside, and wants to do stuff as if it was inside
  - more common, happens when the user wants to make direct event handler calls bypassing `emit()` (and thereby also bypassing our private _hook_is_running) thing.
    This happened to me more than once. I had to go and poke into internal/private stuff, which I shouldn't have to

File: ops/model.py
Comment (PietroPasotti) [reply ops/model.py:951]: I iterated four times on that code and couldn't find any other way to make it clear to myself.
But now that you say it it makes sense. 

*feels old in Python*

Comment (PietroPasotti) [reply ops/model.py:951]: gave it a shot, it's like four times as long now but less nesty. Added a bunch of comments because IMHO it's really not that easy

File: ops/testing.py
Comment (PietroPasotti) [reply ops/testing.py:142]: Conclusion from standup: 
 - for the moment, keep the `event_context` private (--> `_event_context`).
   people won't have a way to enforce 'strict mode' for a while longer
 - open a separate PR where we expose a new `Harness.emit_event(method, event:EventCls, *args, **kwargs)` (or something like that) -- which is the new go-to way to manually fire events with the harness (which internally uses `_event_context` to wrap `emit`.).

**Comment by rwcarlsen** [comment]
Is this ready for a re-review?

**Comment by PietroPasotti** [comment]
> Is this ready for a re-review?

Yes and no. I'm done getting rid of large chunks of the testing data wrappers, but I am left with one big bug which I'm not sure how to tackle best.

Issue is: we want to use `ops.Model` for as much as possible. `ops.Model.relations[idx].data` returns RelationDataContents, which, on `__setitem__`, calls `relation_set`. 
What we want, in order to avoid code duplication and imperfectly replicating model logic in testing code, is to keep that flow as intact as possible.
The issue is that patching `relation_set` in `TestingModelBackend` is not enough, because `relation_set` is plumbed to always be called from the perspective of "this" unit or application, so it won't allow us to set "remote" units/apps databags, as it does not even expose the necessary arguments. So from the perspective of the harness, when we intercept a relation-set call, we can only assume we're setting local databags, which in testing code is not always the case.

Possible solutions: 
- patch `RelationDataContents` and overwrite its `__getitem__/__setitem__` in such a way that `TestingModelBackend.relation_get/relation_set` are called with some extra arguments (namely, `RelationDataContents._entity`). 
   - pros: kind of easy
   - cons: we'd need to still duplicate  `__getitem__/__setitem__`, which contain exactly that validation logic which we were trying to avoid to duplicate.
 - from `TestingModelBackend.relation_get/relation_set`, use `inspect` to get a hold of the outer frame and grab the `_entity` which the relation data is about. (erm...)

**Comment by PietroPasotti** [comment]
> I'm not sure I understand. It is the case that it doesn't work at all (without one of your proposed solutions) or that it doesn't catch and emit errors we want? The test harness itself only represents the "world" from the view of a single charm/unit - so it seems like it should be fine. But maybe I'm not imagining things correct. A concrete example of what we want tests to be able to do and how this problem prevents that might help me.

Right, so it is ready for a re-review.
I chose for the following approach:

add a backend method `update_relation_data` which in 'real' backend does little more than call `relation_set`. The testing backend bypasses `relation_set` and goes directly directly to manipulate the 'raw' relation data.
It works, it feels clean-ish, but I'm wondering now what the purpose of relation-set in the `testingmodelbackend` should be. Should it go?
We get a bunch of broken tests because the calls being recorded have changed, as `relation_set` is now basically out of the picture in the harness. We'd get the same if we were to change the `relation_set` signature directly, so there's no avoiding that. 

Before I fix those, I'd like to get a green light from @pengale as well, and think harder of whether this is the best approach

**Comment by rwcarlsen** [comment]
I've run into those test issues before.  I think it's not ideal to diff internal implementation call stacks for testing :/

**Review by pengale — CHANGES_REQUESTED** [review]
Apologies for not commenting earlier. Here are my thoughts:

This code is based on a careful and clever understanding of the mechanics of the current testing harness and model back end, and I understand how it solves the issues present in the bug.

It does make some pretty big changes to our testing code. The pattern is:

```
def test_something(...):
    setup_task_a
    setup_task_b

    with some_event_context(...):
        exercise_code()

    perform_check_a
    perform_check_b
```

That's an acceptable pattern. And it's a pattern that lets existing tests "just work", albeit without benefiting from more accurate-to-production harness behavior. It is a pretty big shift in the pattern that we're asking people to use, however, and I'm not 100% convinced that wrapping most test code in this particular way is what we want to do.

**Comment by pengale** [comment]
As a follow-up to my comment, I'm not sure that this is the correct solution, but it might be more natural to introduce something like:

```
with as_leader():
    relation.data[app][key] = val
```

You could extend this idea to do other things, like:

```
with as_unit_0():
    set_some_remote_rel_data
```

Or even:

```
with as_human():
    set_some_config_val
```

That way, your decorator only intrudes on the situation when you specifically need it. This approach isn't as friendly to existing tests, which might break if we implement another strict check that better matches production behavior. But it does make stuff a bit more natural to reason about. "For the moment, I'm behaving as the unit that is the leader."

File: ops/framework.py
Comment (rwcarlsen) [inline ops/framework.py:858]: Looks like this needs to be private.

File: ops/model.py
Comment (rwcarlsen) [inline ops/model.py:966]: This feels like it should be more like a `_validate_read` like your already existing `_validate_write`.  It could directly raise the exceptions based on *why* the read permission doesn't exist rather than providing a more generic raise later.  Also, we need to make sure we have high confidence that this doesn't return False (or raise) in any cases where the `relation-get` hook tool would otherwise succeed.

File: ops/model.py
Comment (rwcarlsen) [inline ops/model.py:1006]: It looks like the only usage of `_is_writable` is here.  And so maybe it would be appropriate to just consolidate these two methods - and only have a `_vaidate_write` and a corresponding `_validate_read` that both do all the checks and raise nice, specific exceptions.

**Review by rwcarlsen — CHANGES_REQUESTED** [review]
I'm satisfied with where this ended up.  A couple minor issues are noted below inline. 
 Regarding @pengale's concerns - this PR should only affect testing of ops/harness itself and not affect how charm-writers need to write tests - any breaking of charm tests I would consider to be a bug in this PR that needs to be addressed.  And notably, there were only a couple of small changes necessary to update our internal tests.

File: ops/framework.py
Comment (PietroPasotti) [reply ops/framework.py:858]: Good point, we don't want to risk people touching it from charm code.

**Comment by PietroPasotti** [comment]
I addressed the last two points by @rwcarlsen and left @pengale's comments unaddressed as I tend to agree with Robert's viewpoint.
It's not the goal of this PR to change the way user code is written, its intention was to enable testing code to be more strict, as an opt-in feature, while maintaining compatibility with existing test code.

As a result of this, our testing code had to be changed:
- model tests all need to be wrapped in the 'strict' context to ensure that the validation checks are enabled, because these tests check the framework from the POV of a charm
- testing tests needed little-to-no-changes (so far as the event context is concerned) because they check the framework from the POV of testing code (i.e. code where the access control checks are skipped).

I agree with @pengale that opt-in means: barely anyone will even notice it's there, and that ideally we'd make strict mode the default, but that is a breaking change and out of scope for now.
We could plan ahead a transition period with a warning (like for SIMULATE_CAN_CONNECT), and after a couple of major releases, make strict mode the default.

**Comment by rwcarlsen** [comment]
I'm not even sure that a "strict" mode is preferred as a default.  What is implemented here basically is strict when they (charmers) want it strict (e.g. when running charm code in hooks, etc.) and unrestricted when they don't want restrictions (e.g. when setting up a test, running assertions, etc.).

**Review by rwcarlsen — APPROVED** [review]

**Comment by PietroPasotti** [comment]
> I'm not even sure that a "strict" mode is preferred as a default. What is implemented here basically is strict when they (charmers) want it strict (e.g. when running charm code in hooks, etc.) and unrestricted when they don't want restrictions (e.g. when setting up a test, running assertions, etc.).

Right, maybe the best we can aim for is a good doc explaining how to enable this strict mode when you want to mock an event handler call or so

**Review by pengale — APPROVED** [review]
Apologies for blocking this before. I misunderstood what we were doing. The solution here is complex, but good. +1

**Comment by rwcarlsen** [comment]
@pengale - no need to apologize - your comments were definitely helpful.

**Pull request merged**
